winch: Add a subset of known libcalls and improve call emission

saulecabrera opened the pull request:
This change is a follow up to:
- https://github.com/bytecodealliance/wasmtime/pull/7155
- https://github.com/bytecodealliance/wasmtime/pull/7035

One of the objectives of this change is to make it easy to emit function calls at the MacroAssembler layer, for cases in which it's challenging to know ahead-of-time if a particular functionality can be achieved natively (e.g. rounding and SSE4.2).  The original implementation of function call emission, made this objective difficult to achieve and it was also difficult to reason about.
I decided to simplify the overall approach to function calls as part of this PR; in essence, the `call` module now exposes a single function `FnCall::emit` which is reponsible of gathtering the dependencies and orchestrating the emission of the call. This new approach deliberately avoids holding any state regarding the function call for simplicity.

This change also standardizes the usage of `Callee` as the main entrypoint for function call emission, as of this change 4 `Callee` types exist (`Local`, `Builtin`, `Import`, `FuncRef`), each callee kind is mappable to a `CalleeKind` which is the materialized version of a callee which Cranelift understands.

This change also moves the creation of the `BuiltinFunctions` to the `ISA` level given that they can be safely used accross multiple function compilations.

Finally, this change also introduces support for some of the "well-known" libcalls and hooks those libcalls at the `MacroAssembler::float_round` callsite.

--

prtest:full

<!--
Please make sure you include the following information:

- If this work has been discussed elsewhere, please include a link to that
  conversation. If it was discussed in an issue, just mention "issue #...".

- Explain why this change is needed. If the details are in an issue already,
  this can be brief.

Our development process is documented in the Wasmtime book:
https://docs.wasmtime.dev/contributing-development-process.html

Please ensure all communication follows the code of conduct:
https://github.com/bytecodealliance/wasmtime/blob/main/CODE_OF_CONDUCT.md
-->

saulecabrera commented:
@elliottt this is the change that I mentioned earlier this week during the Cranelift weekly meeting.

fitzgen commented on winch/codegen/src/codegen/env.rs:
```diff
@@ -32,17 +29,46 @@ pub struct TableData {
+    ///
+    /// # Panics
+    /// This function panics if the [`Callee`] is not a built-in function.
+    pub fn get_builtin(&self) -> &BuiltinFunction {
```
Maybe name this `unwrap_builtin`?

saulecabrera replied:
As I was fixing this I realized that this was actually dead code, so I ended up removing `get_builtin` and `get_info` entirely.

fitzgen commented on winch/codegen/src/codegen/env.rs:
```diff
@@ -54,15 +80,13 @@ pub struct CalleeInfo {
 /// Contains all information about the module and runtime that is accessible to
 /// to a particular function during code generation.
-pub struct FuncEnv<'a, P: PtrSize> {
+pub struct FuncEnv<'a, 'b: 'a, 'c: 'b, P: PtrSize> {
```
In my experience, when we start to get many lifetime parameters like this, it helps to give them more descriptive names.

fitzgen commented on winch/codegen/src/codegen/mod.rs:
```diff
@@ -22,21 +18,21 @@ pub(crate) use call::*;
 
 /// The code generation abstraction.
-pub(crate) struct CodeGen<'a, M>
+pub(crate) struct CodeGen<'a, 'b: 'a, 'c: 'b, M>
```
ditto

saulecabrera commented on winch/codegen/src/codegen/builtin.rs:
```diff
@@ -5,16 +5,57 @@ use crate::{
+pub(crate) enum BuiltinType {
+    /// Dynamic built-in function, derived from the VMContext.
+    Dynamic {
+        /// The offset of the built-in function.
+        offset: u32,
```
Yeah, but note that this index is the built-in function's index multiplied by the pointer size, but index works for me.

saulecabrera commented on winch/codegen/src/codegen/builtin.rs:
```diff
@@ -5,16 +5,57 @@ use crate::{
+    Dynamic {
+        /// The offset of the built-in function.
+        offset: u32,
+        /// The built-in function base, relative to the VMContext.
+        base: u32,
```
I don't think that this will work (unless I'm missing something). The reason why I'm storing both separately is that loading the builtin's address requires two loads:
- base = load(vmctx + base)
- builtin_addr = load(base + index)

elliottt commented on winch/codegen/src/codegen/env.rs:
```diff
@@ -54,15 +55,13 @@ pub struct CalleeInfo {
 /// Contains all information about the module and runtime that is accessible to
 /// to a particular function during code generation.
-pub struct FuncEnv<'a, P: PtrSize> {
+pub struct FuncEnv<'a, 'translation: 'a, 'data: 'translation, P: PtrSize> {
```
Similarly to the `CodeGenContext` parameters, can this be simplified down to using the `'a` lifetime for everything?

saulecabrera replied:
Yeah, [the same comment](https://github.com/bytecodealliance/wasmtime/pull/7228#discussion_r1358321683) applies here too.

elliottt reviewed (commented):
This is looking great, I'm so happy to see the default cases for fp rounding operations now!

saulecabrera commented on winch/codegen/src/codegen/context.rs:
```diff
@@ -27,25 +28,37 @@ use std::ops::RangeBounds;
 /// be generally used as the single entry point to access
 /// the compound functionality provided by its elements.
-pub(crate) struct CodeGenContext<'a> {
+pub(crate) struct CodeGenContext<'a, 'b: 'a> {
```
It actually outlives 'a; the changes in this PR make it so the `compiled_function` function at the ISA level gets a reference to `BuiltinFunctions`, which will outlive any of the other structs created in the body of the `compile_function` function (e.g. `FuncEnv` and `VMOffsets`) and on top of that, there's is a third lifetime ('`data`, previously `'c`) which represents the lifetime of any references to the original WebAssembly binary through `wasmtime_environ::ModuleTranslation`.

saulecabrera commented:
I've addressed all of the review comments (thanks @fitzgen and @elliottt). I'll go ahead and merge this, but happy to follow up on any of the comments if need be.
